### PR TITLE
HBW-146 Dynamic disable_if and delete_if options for string

### PR DESCRIPTION
**Pull request by sugarfree1**

### CODE REVIEW

File: hbw/app/javascript/packs/hbw/components/shared/hoc/conditions.js [hunk 1]
```diff
@@ -61,11 +69,16 @@ export default WrappedComponent => class WithConditions extends Component {
 
   some = results => results.some(el => el === true);
 
+  fireFieldValueUpdate = (name, value) => {
+    this.props.trigger('hbw:update-value', { name, value });
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/shared/hoc/conditions.js:73]: Does it work properly with several fields subscribed on the same variable?

File: hbw/config/yml_api.test.camunda.yml [hunk 2]
```diff
@@ -856,6 +862,122 @@ get:
...
+            label: Dynamic disable_if test
+            css_class: col-xs-12
+            fields:
+              - name: homsOrderCheckingThis
+                css_class: col-xs-6 col-sm-4 col-md-3
+                label: Reason
+                type: string
+              - name: homsOrderOrEnableAfterChanges
+                type: string
+                label: Will be enable after changes
```
Comment (TimAle) [inline hbw/config/yml_api.test.camunda.yml:881]:
```suggestion
                label: Will be enabled after changes
```
and so on

File: spec/hbw/features/bp_form_dynamic_disable_fields_spec.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    click_and_wait 'ORD-25'
+
+    expect(page).to have_selector "[name='homsOrderCheckingThis']"
+  end
+
+  scenario 'AND & OR conditions for string' do
+    expect(find_by_name('homsOrderAndEnableAfterChanges').readonly?).to  eq true
+    expect(find_by_name('homsOrderAndDisableAfterChanges').readonly?).to eq false
+    expect(find_by_name('homsOrderOrEnableAfterChanges').readonly?).to   eq true
+    expect(find_by_name('homsOrderOrDisableAfterChanges').readonly?).to  eq false
```
Comment (TimAle) [inline spec/hbw/features/bp_form_dynamic_disable_fields_spec.rb:25]: Maybe we can call some helpers for that?

File: spec/hbw/features/bp_form_dynamic_disable_fields_spec.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    expect(find_by_name('homsOrderAndEnableAfterChanges').readonly?).to  eq false
+    expect(find_by_name('homsOrderAndDisableAfterChanges').readonly?).to eq true
+    expect(find_by_name('homsOrderOrEnableAfterChanges').readonly?).to   eq false
+    expect(find_by_name('homsOrderOrDisableAfterChanges').readonly?).to  eq true
+
+    expect(page).to have_selector "button[type='submit']"
+    click_and_wait 'Submit'
+
+    expect_widget_presence
```
Comment (TimAle) [inline spec/hbw/features/bp_form_dynamic_disable_fields_spec.rb:37]: Why should we submit this?

**Review by TimAle — CHANGES_REQUESTED** [review]

File: hbw/app/javascript/packs/hbw/components/shared/hoc/conditions.js (hunk 1 above)
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/shared/hoc/conditions.js:73]: form
"1" - one field disabled, "2" - two fields disabled, "3" - group disabled

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,9 +1,18 @@
...
+### Changes
+Change serialization of disable, read only and hidden fields. Now all values of these fields will not be submitted.
+See [#296](https://github.com/latera/homs/pull/296) for more details.
+
 ### Features
 - [#493](https://github.com/latera/homs/pull/493) Add missing labels for inputs on search order form.
 - [#497](https://github.com/latera/homs/pull/497) Handle connection error when try to print order.
 - [#499](https://github.com/latera/homs/pull/499) Increase memory limit for worker.
+- [#289](https://github.com/latera/homs/pull/289) Make conditions `disable_if` and `delete_if` dynamic.
+- [#294](https://github.com/latera/homs/pull/294) Add test for dynamic disable and delete.
```
[the remark this answers is not in the document]
Comment (vladdoncheg) [reply CHANGELOG.md:12]: Nope. It was a legacy :)

File: hbw/app/javascript/packs/hbw/components/entity_task.js.jsx [hunk 6]
```diff
@@ -71,7 +71,8 @@ modulejs.define(
               env:               this.props.env,
               error:             this.state.error,
               assignee:          this.props.task.assignee,
-              variables:         this.formVariablesFromTask(this.props.task)
+              variables:         this.formVariablesFromTask(this.props.task),
+              taskVariables:     this.props.task.form.variables
```
[the remark this answers is not in the document]
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/entity_task.js.jsx:75]: Done

File: hbw/app/javascript/packs/hbw/components/form/file_upload.js.jsx [hunk 7]
```diff
@@ -128,7 +128,13 @@ modulejs.define('HBWFormFileUpload', ['React'], (React) => {
       return fileReader.readAsBinaryString(file);
     };
 
-    serialize = () => ({ [this.props.params.name]: JSON.stringify({ files: this.state.files }) });
+    serialize = () => {
+      if (this.props.hidden) {
```
[the remark this answers is not in the document]
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/form/file_upload.js.jsx:132]: Done

File: hbw/config/yml_api.test.camunda.yml (hunk 2 above)
Comment (vladdoncheg) [reply hbw/config/yml_api.test.camunda.yml:881]: Done

File: spec/hbw/features/bp_form_dynamic_disable_fields_spec.rb (hunk 4 above)
Comment (vladdoncheg) [reply spec/hbw/features/bp_form_dynamic_disable_fields_spec.rb:37]: To check that hidden and disabled fields were not submitted, I think

File: spec/hbw/features/bp_form_dynamic_disable_fields_spec.rb (hunk 3 above)
Comment (vladdoncheg) [reply spec/hbw/features/bp_form_dynamic_disable_fields_spec.rb:25]: Done

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 8]
```diff
@@ -200,8 +204,22 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+
     notSerializableFields = () => ['static'];
 
+    setInitialFormValues = () => {
+      const variables = {};
+
+      this.props.taskVariables.forEach((variable) => {
+        variables[variable.name] = variable.value;
+      });
+
```
[the remark this answers is not in the document]
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/form.js.jsx:219]: Yep,  done

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 9]
```diff
@@ -15,15 +15,18 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
     };
 
     componentDidMount () {
+      this.setInitialFormValues();
       jQuery(':input:enabled:visible:first').focus();
       this.props.bind(`hbw:submit-form-${this.props.id}`, () => this.setState({ submitting: true }));
       this.props.bind('hbw:form-submitting-failed', () => this.setState({ submitting: false }));
       this.props.bind('hbw:file-upload-started', () => this.setState({ fileUploading: true }));
       this.props.bind('hbw:file-upload-finished', () => this.setState({ fileUploading: false }));
+      this.props.bind('hbw:update-value', data => this.updateFormValues(data));
```
[the remark this answers is not in the document]
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/form.js.jsx]: Added id, now it works correctly

**Comment by vladdoncheg** [comment]
Finally we have:
* tests to check that delete_if and disable_if work with AND and OR;
* test to check that all type of fields could dependent on String;
* tests to check that hidden and disabled fields will not be submitted.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,9 +1,17 @@
 v2.5.0 [unreleased]
 -------------------
+### Changes
+Change serialization of disable, read only and hidden fields. Now all values of these fields will not be submitted.
```
Comment (TimAle) [inline CHANGELOG.md:4]:
```suggestion
Change serialization of disabled, read-only and hidden fields. Now all values of these fields will not be submitted.
```

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 11]
```diff
@@ -86,7 +89,8 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
         value:           variables[name],
         formSubmitting:  submitting || fileUploading,
         fileListPresent: this.fileListPresent(form.fields),
-        showSubmit:      !!assignee
+        showSubmit:      !!assignee,
+        formValues:      this.state.formValues
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:93]: Prefetch it like other state and props.

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 12]
```diff
@@ -200,8 +204,21 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+      this.setState(prevState => ({ formValues: { ...prevState.formValues, [name]: value } }));
+    };
+
     notSerializableFields = () => ['static'];
 
+    setInitialFormValues = () => {
+      const variables = this.props.taskVariables.reduce((result, variable) => {
+        result[variable.name] = variable.value;
+        return result;
+      }, {});
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:217]:
```suggestion
      const variables = this.props.taskVariables.reduce((result, { name, value }) => {
        result[name] = value;
        return result;
      }, {});
```

File: hbw/app/javascript/packs/hbw/components/form/group.js.jsx [hunk 13]
```diff
@@ -123,6 +124,10 @@ modulejs.define('HBWFormGroup', ['React', 'HBWFormDatetime',
     notSerializableFields = () => ['static'];
 
     serialize = () => {
+      if (this.props.hidden || this.props.disabled) {
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form/group.js.jsx:127]: `hidden` prop only.
`disabled` prop is delegated to nested components.

File: spec/hbw/features/bp_form/dynamic_conditions_by_string_spec.rb [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    expect(page).to     have_content 'Orders list'
+
+    order_type = FactoryBot.create(:order_type, :support_request)
+
+    FactoryBot.create(:order, order_type: order_type).update(code: 'ORD-27')
+
+    click_on 'Orders'
+    expect(page).to have_content 'Orders list'
+    expect_widget_presence
+
```
Comment (TimAle) [inline spec/hbw/features/bp_form/dynamic_conditions_by_string_spec.rb:15]: There is no need for this.
Just create your order before signing in.

File: spec/hbw/features/bp_form/dynamic_conditions_by_string_spec.rb [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    expect(page).to have_content  'Dependent static'
+    expect(page).to have_selector '.dependent-select'
+    expect(page).to have_selector '.dependent-chechbox'
+    expect(page).to have_selector '.dependent-select-table'
+    expect(page).to have_selector '.dependent-file-list'
+    expect(page).to have_selector '.dependent-group'
+    expect(page).to have_selector "[name='dependentDatetime']"
+    expect(page).to have_selector "[name='dependentString']"
+    expect(page).to have_selector "[name='dependentText']"
+    expect(page).to have_selector "[name='dependentUploadedFile']"
```
Comment (TimAle) [inline spec/hbw/features/bp_form/dynamic_conditions_by_string_spec.rb:31]: Why not by labels?

File: spec/hbw/features/bp_form/dynamic_disabled_fields_spec.rb [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    click_on 'Orders'
+    expect(page).to have_content 'Orders list'
+    expect_widget_presence
+
+    click_and_wait 'ORD-25'
+
+    expect(page).to have_selector "[name='homsOrderCheckingThis']"
+  end
+
+  scenario 'disable for AND & OR conditions' do
```
Comment (TimAle) [inline spec/hbw/features/bp_form/dynamic_disabled_fields_spec.rb:21]: `disabled` or `should be disabled`

File: spec/hbw/features/bp_form/dynamic_disabled_fields_spec.rb [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    expect(readonly?('homsOrderAndEnabledAfterChanges')).to  eq true
```
Comment (TimAle) [inline spec/hbw/features/bp_form/dynamic_disabled_fields_spec.rb:22]: Let's prefer `be` matcher with booleans.

File: spec/hbw/features/bp_form/submit_disabled_fields_spec.rb [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    click_on 'Orders'
+    expect(page).to have_content 'Orders list'
+    expect_widget_presence
+
+    click_and_wait 'ORD-28'
+
+    expect(page).to have_selector "[name='submittedString']"
+  end
+
+  scenario 'do not submit disabled fields' do
```
Comment (TimAle) [inline spec/hbw/features/bp_form/submit_disabled_fields_spec.rb:21]: Do something with example names, please.

**Review by TimAle — CHANGES_REQUESTED** [review]

File: spec/hbw/features/bp_form/dynamic_disabled_fields_spec.rb (hunk 16 above)
Comment (vladdoncheg) [reply spec/hbw/features/bp_form/dynamic_disabled_fields_spec.rb:21]: Done

File: hbw/app/javascript/packs/hbw/components/form/group.js.jsx (hunk 13 above)
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/form/group.js.jsx:127]: Done

File: hbw/app/javascript/packs/hbw/components/form.js.jsx (hunk 12 above)
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/form.js.jsx:217]: Done

File: hbw/app/javascript/packs/hbw/components/form.js.jsx (hunk 11 above)
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/form.js.jsx:93]: Done

File: CHANGELOG.md (hunk 10 above)
Comment (vladdoncheg) [reply CHANGELOG.md:4]: Done

File: spec/hbw/features/bp_form/submit_disabled_fields_spec.rb (hunk 18 above)
Comment (vladdoncheg) [reply spec/hbw/features/bp_form/submit_disabled_fields_spec.rb:21]: Done

File: hbw/app/javascript/packs/hbw/components/form/string.js.jsx [hunk 19]
```diff
@@ -374,7 +376,13 @@ modulejs.define('HBWFormString', ['React'], (React) => {
       }
     };
 
-    serialize = () => ({ [this.props.name]: this.state.value });
+    serialize = () => {
+      if (this.props.params.editable === false || this.props.disabled || this.props.hidden) {
```
Comment (sugarfree1) [inline hbw/app/javascript/packs/hbw/components/form/string.js.jsx:380]: that condition looks like a helper, you duplicate that check in every field

**Pull request merged**